# Homework 01: Data Cleaning and Exploratory Data Analysis 
***

**Name**: Matthew Riopelle

***

This assignment is due on Canvas by **6:00PM on Friday September 2**. Your solutions to theoretical questions should be done in Markdown directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  Remember that you are encouraged to discuss the problems with your classmates, but **you must write all code and solutions on your own**.

**NOTES**: 

- Any relevant data sets should be available in the Homework 01 assignment write-up on Canvas. To make life easier on the grader if they need to run your code, do not change the relative path names here. Instead, move the files around on your computer.
- If you're not familiar with typesetting math directly into Markdown then by all means, do your work on paper first and then typeset it later.  Remember that there is a [reference guide](https://math.meta.stackexchange.com/questions/5020/mathjax-basic-tutorial-and-quick-reference) linked on Canvas on writing math in Markdown. **All** of your written commentary, justifications and mathematical work should be in Markdown.
- Because you can technically evaluate notebook cells is a non-linear order, it's a good idea to do $\color{red}{\text{Kernel}}$ $\color{red}\rightarrow$ $\color{red}{\text{Restart & Run All}}$ as a check before submitting your solutions.  That way if we need to run your code you will know that it will work as expected. 
- It is **bad form** to make your reader interpret numerical output from your code.  If a question asks you to compute some value from the data you should show your code output **AND** $\color{red}{\text{write a summary of the results}}$ in Markdown directly below your code. 
- This probably goes without saying, but... For any question that asks you to calculate something, you **must show all work and justify your answers to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit. 

---

Per usual you should import Pandas and NumPy when you start working with data.

In [1]:
import pandas as pd
import numpy as np


Also, lets read in the csv file. The path you use should reflect where you are keeping the data. For simplicity sake, put the data file (i.e. the .csv file) in the same folder as the Jupyter notebook file.

You will create a dataframe (df) called **CollegeType**. The standard is to precede the name with `df`, therefore call it `dfCollegeType` and read it in.


One way to do this (not recommended) is required when you store the data in a different location than the Jupyter file. See below.

The code is commented out; don't uncomment the cell and run it unless you are actually keeping your data in an alternate location than your Jupyter file for HW1.

In [2]:
# dfCollegeType = pd.read_csv("/location1/location2/location3/.../salaries-by-college-type.csv")

Now, if you are actually keeping the data in the same folder as the Jupyter file, we can read in the csv file with the following:

In [3]:
dfCollegeType = pd.read_csv("salaries-by-college-type.csv")

# Problem 1
***
Lets look at the data you just read in to see how it appears, and decide how it needs cleaned (all data needs cleaned).

$\color{blue}{\text{FYI, in the future always look over your data and clean it if it hasn't already been cleaned.}}$

In [4]:
dfCollegeType

School Name  School Type  \
0     Massachusetts Institute of Technology (MIT)  Engineering   
1        California Institute of Technology (CIT)  Engineering   
2                             Harvey Mudd College  Engineering   
3    Polytechnic University of New York, Brooklyn  Engineering   
4                                    Cooper Union  Engineering   
..                                            ...          ...   
264                  Austin Peay State University        State   
265                    Pittsburg State University        State   
266                      Southern Utah University        State   
267           Montana State University - Billings        State   
268                  Black Hills State University        State   

    Starting Median Salary Mid-Career Median Salary  \
0              $72,200.00              $126,000.00    
1              $75,500.00              $123,000.00    
2              $71,800.00              $122,000.00    
3              $62,400.00              $114,000.00    
4              $62,200.00              $114,000.00    
..                     ...                      ...   
264            $37,700.00               $59,200.00    
265            $40,400.00               $58,200.00    
266            $41,900.00               $56,500.00    
267            $37,900.00               $50,600.00    
268            $35,300.00               $43,900.00    

    Mid-Career 10th Percentile Salary Mid-Career 25th Percentile Salary  \
0                         $76,800.00                        $99,200.00    
1                                 NaN                      $104,000.00    
2                                 NaN                       $96,000.00    
3                         $66,800.00                        $94,300.00    
4                                 NaN                       $80,200.00    
..                                ...                               ...   
264                       $32,200.00                        $40,500.00    
265                       $25,600.00                        $46,000.00    
266                       $30,700.00                        $39,700.00    
267                       $22,600.00                        $31,800.00    
268                       $27,000.00                        $32,200.00    

    Mid-Career 75th Percentile Salary Mid-Career 90th Percentile Salary  
0                        $168,000.00                       $220,000.00   
1                        $161,000.00                                NaN  
2                        $180,000.00                                NaN  
3                        $143,000.00                       $190,000.00   
4                        $142,000.00                                NaN  
..                                ...                               ...  
264                       $73,900.00                        $96,200.00   
265                       $84,600.00                       $117,000.00   
266                       $78,400.00                       $116,000.00   
267                       $78,500.00                        $98,900.00   
268                       $60,900.00                        $87,600.00   

[269 rows x 8 columns]

# Description of this data set
This data is a collection of median salaries from 269 institutions of various types.

The median salary was calculated at various times (Starting salary and mid-career) in the careers of graduates from these same institutions . For a particular year, each school reported the starting and mid-career salary for approximately 100 students. Each group of 100 was chosen with a random number generator from a list of alumni for that particular school.

Therefore, each school can list the median **starting salary** as well as the median **mid-career salary**.
Also, along with the median (50th percentile) some schools also reported the 10th, the 25th, the 75th and the 90th percentile mid-career salaries .

Notice that some schools did not report all percentiles, hence the $\color{red}{\text{NAN's}}$ seen in the data.


# Part A
### Answer the following questions:

1] ***(2 points)*** What is the $\color{red}{\text{VOI}}$ (Variable of Interest) in this data set?

The variable of interest in this data set is the median salaries of graduates

2] ***(2 points)*** What is the $\color{red}{\text{sample frame}}$ for this data set?

The sample frame would be educational institutions in America

3] ***(2 points)*** What $\color{red}{\text{type of sample}}$ is this (convenience sample,systematic sample, census sample, stratified sample,...) ?

It states that the collections is of 269 institutions of various types; however, not much more information is given. Since the schools are laid out by type, I would have to guess that the sample type would be stratified, which is when a heterogenous population is broken into homogenous groups, then sampled randomly. So, for example, they would bunch together state schools in one group, engineering schools in another, etc., then sample from those groups.

4] ***(2 points)*** What is the $\color{red}{\text{population}}$ and what is the $\color{red}{\text{sample}}$?

The population of this dataset would be the graduates of those educational institutions. The sample itself would be the salaries of those within that population.

## Example help for cleaning (Required in Part B)

Now that we have an idea of what this data is and where it came from, we might want to construct some questions of interest like, "What type of school has the highest starting salary?"

However, before we can work with the data, it needs to be $\color{red}{\text{cleaned}}$!

For starters, notice the numerical columns are strings with non-numerical symbols such as: `$`, `,`, `.`, and `NaN`

In **general**, here is a way to clean data similar to this set.
 All the string surgery is seen in the next cell:

In [5]:
#df[col] = df[col].replace('$', '')  # assuming there exist '$' symbols in the data
#df[col] = df[col].replace(',', '')  # assuming ',' is the thousand's separator in the data
#df[col] = df[col].replace('%', '')  # assuming there exist '%' symbols in the data
## etc. for any other undesired symbols in the data

Then we would cast the data in the column as a float, as seen in the next cell:

In [6]:
#df[col] = df[col].astype(float)  # cast back to appropriate type

**Specifically**, you should run the following code for the column entitled '$\color{blue}{\text{Starting Median Salary}}$' in this data set:

In [7]:
# Eliminate the ',' from the strings in the 'Starting Median Salary' column
dfCollegeType['Starting Median Salary'] = dfCollegeType['Starting Median Salary'].str.replace(',','')

# Eliminate the '$' from the strings in the 'Startin Median Salary' column
dfCollegeType['Starting Median Salary'] = dfCollegeType['Starting Median Salary'].str.replace('$','')

# Cast the strings in the 'Starting Median Salary' as floats.
dfCollegeType['Starting Median Salary'] = dfCollegeType['Starting Median Salary'].astype(float)

# Now take a look at the dataframe. The strings in the column 'Starting Median Salary' should be floats.
dfCollegeType


/tmp/ipykernel_378/72921303.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  dfCollegeType['Starting Median Salary'] = dfCollegeType['Starting Median Salary'].str.replace('$','')


School Name  School Type  \
0     Massachusetts Institute of Technology (MIT)  Engineering   
1        California Institute of Technology (CIT)  Engineering   
2                             Harvey Mudd College  Engineering   
3    Polytechnic University of New York, Brooklyn  Engineering   
4                                    Cooper Union  Engineering   
..                                            ...          ...   
264                  Austin Peay State University        State   
265                    Pittsburg State University        State   
266                      Southern Utah University        State   
267           Montana State University - Billings        State   
268                  Black Hills State University        State   

     Starting Median Salary Mid-Career Median Salary  \
0                   72200.0             $126,000.00    
1                   75500.0             $123,000.00    
2                   71800.0             $122,000.00    
3                   62400.0             $114,000.00    
4                   62200.0             $114,000.00    
..                      ...                      ...   
264                 37700.0              $59,200.00    
265                 40400.0              $58,200.00    
266                 41900.0              $56,500.00    
267                 37900.0              $50,600.00    
268                 35300.0              $43,900.00    

    Mid-Career 10th Percentile Salary Mid-Career 25th Percentile Salary  \
0                         $76,800.00                        $99,200.00    
1                                 NaN                      $104,000.00    
2                                 NaN                       $96,000.00    
3                         $66,800.00                        $94,300.00    
4                                 NaN                       $80,200.00    
..                                ...                               ...   
264                       $32,200.00                        $40,500.00    
265                       $25,600.00                        $46,000.00    
266                       $30,700.00                        $39,700.00    
267                       $22,600.00                        $31,800.00    
268                       $27,000.00                        $32,200.00    

    Mid-Career 75th Percentile Salary Mid-Career 90th Percentile Salary  
0                        $168,000.00                       $220,000.00   
1                        $161,000.00                                NaN  
2                        $180,000.00                                NaN  
3                        $143,000.00                       $190,000.00   
4                        $142,000.00                                NaN  
..                                ...                               ...  
264                       $73,900.00                        $96,200.00   
265                       $84,600.00                       $117,000.00   
266                       $78,400.00                       $116,000.00   
267                       $78,500.00                        $98,900.00   
268                       $60,900.00                        $87,600.00   

[269 rows x 8 columns]

# Part B
# Find the mean of the column 'Starting Median Salary' by hand.

The data in the column 'Starting Median Salary' is now "clean" in that calculations can be performed on the floats held in that column.

Let find the mean of this column by performing the hand calculation $\displaystyle\bar{x}=\frac{1}{n}\cdot\sum_{i=1}^nx_i$

Sum the numbers in the column and store them in a variable called `sum_xs`

`sum_xs = dfCollegeType['Starting Median Salary'].sum()`

Find out how many items are in the data frame and store that number in a variable called `n`

`n = len(dfCollegeType.index)`

***(2 points)*** Create a variable `mean_SMS_byHand`. Perform the required calculation for a mean and print it to the screen.


In [8]:
# code here for 2 points
sum_xs = dfCollegeType['Starting Median Salary'].sum()
n = len(dfCollegeType.index)
#calculating mean below
mean_SMS_byHand = sum_xs/n
print("The mean of the 'Starting Median Salary' column is {}".format(mean_SMS_byHand))

The mean of the 'Starting Median Salary' column is 46068.40148698885


### Example help for Parts C and D

Now, lets find that same mean again by simply using the packaged `.mean` function.

Create variable 'mean_SMS' to hold the mean of the column 'Starting Median Salary'

You may want to notice the arguments for the `.mean` function:

`axis = 0` implies the **column** of data as opposed to the **row** of data. `axis = 1` implies row.
`skipna` means to skip any data with an 'NAN' in it.

`axis = 0` is not necessary in this situation and `skipna = True` is the default and also need not be put here.



In [9]:
# We now put the mean of the column 'Starting Median Salary' into the variable `mean_SMS`

# mean_SMS = dfCollegeType[['Starting Median Salary']].mean(axis = 0, skipna = True)
# or
mean_SMS = dfCollegeType[['Starting Median Salary']].mean()

# Then we print 'mean_SMS' to the screen
print(mean_SMS)


Starting Median Salary    46068.401487
dtype: float64


Let's clean the data for the 'Mid-Career 10th Percentile Salary' column.

In [10]:
dfCollegeType['Mid-Career 10th Percentile Salary'] = dfCollegeType['Mid-Career 10th Percentile Salary'].str.replace(',','')
dfCollegeType['Mid-Career 10th Percentile Salary'] = dfCollegeType['Mid-Career 10th Percentile Salary'].str.replace('$','')
dfCollegeType['Mid-Career 10th Percentile Salary'] = dfCollegeType['Mid-Career 10th Percentile Salary'].astype(float)

# Now take a look at the dataframe. The strings in the column 'Mid-Career 25th Percentile Salary' should be floats.
dfCollegeType



/tmp/ipykernel_378/3970973073.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  dfCollegeType['Mid-Career 10th Percentile Salary'] = dfCollegeType['Mid-Career 10th Percentile Salary'].str.replace('$','')


School Name  School Type  \
0     Massachusetts Institute of Technology (MIT)  Engineering   
1        California Institute of Technology (CIT)  Engineering   
2                             Harvey Mudd College  Engineering   
3    Polytechnic University of New York, Brooklyn  Engineering   
4                                    Cooper Union  Engineering   
..                                            ...          ...   
264                  Austin Peay State University        State   
265                    Pittsburg State University        State   
266                      Southern Utah University        State   
267           Montana State University - Billings        State   
268                  Black Hills State University        State   

     Starting Median Salary Mid-Career Median Salary  \
0                   72200.0             $126,000.00    
1                   75500.0             $123,000.00    
2                   71800.0             $122,000.00    
3                   62400.0             $114,000.00    
4                   62200.0             $114,000.00    
..                      ...                      ...   
264                 37700.0              $59,200.00    
265                 40400.0              $58,200.00    
266                 41900.0              $56,500.00    
267                 37900.0              $50,600.00    
268                 35300.0              $43,900.00    

     Mid-Career 10th Percentile Salary Mid-Career 25th Percentile Salary  \
0                              76800.0                       $99,200.00    
1                                  NaN                      $104,000.00    
2                                  NaN                       $96,000.00    
3                              66800.0                       $94,300.00    
4                                  NaN                       $80,200.00    
..                                 ...                               ...   
264                            32200.0                       $40,500.00    
265                            25600.0                       $46,000.00    
266                            30700.0                       $39,700.00    
267                            22600.0                       $31,800.00    
268                            27000.0                       $32,200.00    

    Mid-Career 75th Percentile Salary Mid-Career 90th Percentile Salary  
0                        $168,000.00                       $220,000.00   
1                        $161,000.00                                NaN  
2                        $180,000.00                                NaN  
3                        $143,000.00                       $190,000.00   
4                        $142,000.00                                NaN  
..                                ...                               ...  
264                       $73,900.00                        $96,200.00   
265                       $84,600.00                       $117,000.00   
266                       $78,400.00                       $116,000.00   
267                       $78,500.00                        $98,900.00   
268                       $60,900.00                        $87,600.00   

[269 rows x 8 columns]

**And finally lets print the mean for `Mid-Career 10th Percentile Salary` column**

In [11]:
mean_MC10 = dfCollegeType[['Mid-Career 10th Percentile Salary']].mean(axis = 0, skipna = True)
print(mean_MC10)

Mid-Career 10th Percentile Salary    44250.649351
dtype: float64


# Part C
# Clean the data for the following three columns:
   
   1] ***(3 points)*** Clean the data in column 'Mid-Career Median Salary'
   
   2] ***(3 points)*** Clean the data in column 'Mid-Career 25th Percentile Salary'
   
   3] ***(3 points)*** Clean the data in column 'Mid-Career 75th Percentile Salary'
   

In [12]:
# Put your code here for Part C

#Mid-Career Median Salary Cleaning
dfCollegeType['Mid-Career Median Salary'] = dfCollegeType['Mid-Career Median Salary'].str.replace(',','')
dfCollegeType['Mid-Career Median Salary'] = dfCollegeType['Mid-Career Median Salary'].str.replace('$','')
dfCollegeType['Mid-Career Median Salary'] = dfCollegeType['Mid-Career Median Salary'].astype(float)

#Mid-Career 25th Percentile Salary Cleaning
dfCollegeType['Mid-Career 25th Percentile Salary'] = dfCollegeType['Mid-Career 25th Percentile Salary'].str.replace(',','')
dfCollegeType['Mid-Career 25th Percentile Salary'] = dfCollegeType['Mid-Career 25th Percentile Salary'].str.replace('$','')
dfCollegeType['Mid-Career 25th Percentile Salary'] = dfCollegeType['Mid-Career 25th Percentile Salary'].astype(float)

#Mid-Career 75th Percentile Salary
dfCollegeType['Mid-Career 75th Percentile Salary'] = dfCollegeType['Mid-Career 75th Percentile Salary'].str.replace(',','')
dfCollegeType['Mid-Career 75th Percentile Salary'] = dfCollegeType['Mid-Career 75th Percentile Salary'].str.replace('$','')
dfCollegeType['Mid-Career 75th Percentile Salary'] = dfCollegeType['Mid-Career 75th Percentile Salary'].astype(float)

dfCollegeType

/tmp/ipykernel_378/2596064836.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  dfCollegeType['Mid-Career Median Salary'] = dfCollegeType['Mid-Career Median Salary'].str.replace('$','')
/tmp/ipykernel_378/2596064836.py:10: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  dfCollegeType['Mid-Career 25th Percentile Salary'] = dfCollegeType['Mid-Career 25th Percentile Salary'].str.replace('$','')
/tmp/ipykernel_378/2596064836.py:15: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  dfCollegeType['Mid-Career 75th Percentile Salary

School Name  School Type  \
0     Massachusetts Institute of Technology (MIT)  Engineering   
1        California Institute of Technology (CIT)  Engineering   
2                             Harvey Mudd College  Engineering   
3    Polytechnic University of New York, Brooklyn  Engineering   
4                                    Cooper Union  Engineering   
..                                            ...          ...   
264                  Austin Peay State University        State   
265                    Pittsburg State University        State   
266                      Southern Utah University        State   
267           Montana State University - Billings        State   
268                  Black Hills State University        State   

     Starting Median Salary  Mid-Career Median Salary  \
0                   72200.0                  126000.0   
1                   75500.0                  123000.0   
2                   71800.0                  122000.0   
3                   62400.0                  114000.0   
4                   62200.0                  114000.0   
..                      ...                       ...   
264                 37700.0                   59200.0   
265                 40400.0                   58200.0   
266                 41900.0                   56500.0   
267                 37900.0                   50600.0   
268                 35300.0                   43900.0   

     Mid-Career 10th Percentile Salary  Mid-Career 25th Percentile Salary  \
0                              76800.0                            99200.0   
1                                  NaN                           104000.0   
2                                  NaN                            96000.0   
3                              66800.0                            94300.0   
4                                  NaN                            80200.0   
..                                 ...                                ...   
264                            32200.0                            40500.0   
265                            25600.0                            46000.0   
266                            30700.0                            39700.0   
267                            22600.0                            31800.0   
268                            27000.0                            32200.0   

     Mid-Career 75th Percentile Salary Mid-Career 90th Percentile Salary  
0                             168000.0                      $220,000.00   
1                             161000.0                               NaN  
2                             180000.0                               NaN  
3                             143000.0                      $190,000.00   
4                             142000.0                               NaN  
..                                 ...                               ...  
264                            73900.0                       $96,200.00   
265                            84600.0                      $117,000.00   
266                            78400.0                      $116,000.00   
267                            78500.0                       $98,900.00   
268                            60900.0                       $87,600.00   

[269 rows x 8 columns]

# Part D
# Find the means for the same columns cleaned in part C.
    Variable names should be:
       
   1]  ***(2 points)*** `mean_MCM` for the mean of 'Mid-Career Median Salary' column values.
   
   2]  ***(2 points)*** `mean_MC25` for the mean of 'Mid-Career 25th Percentile Salary' column values.
   
   3]  ***(2 points)*** `mean_MC75` for the mean of 'Mid-Career 75th Percentile Salary' column values.
    

In [13]:
# Put your code here for Part D
#mid career
mean_MCM = dfCollegeType[['Mid-Career Median Salary']].mean(axis = 0, skipna = True)
print(mean_MCM)
#mid career 25th percentile
mean_MC25 = dfCollegeType[['Mid-Career 25th Percentile Salary']].mean(axis = 0, skipna = True)
print(mean_MC25)
#mid career 75th percentile
mean_MC75 = dfCollegeType[['Mid-Career 75th Percentile Salary']].mean(axis = 0, skipna = True)
print(mean_MC75)

Mid-Career Median Salary    83932.342007
dtype: float64
Mid-Career 25th Percentile Salary    60373.234201
dtype: float64
Mid-Career 75th Percentile Salary    116275.092937
dtype: float64


# Problem 2
***
So now we have a single number (mean) that describes the center of the data provided in each column. But what technically did we just find?

**Answer:** The mean of $\color{red}{\text{all}}$ schools' median salaries; either starting or mid-career. This may not be of the utmost interest to us though since all schools are lumped into this one number.

Let's instead seperate the types of institutions to find out what type of school has the highest starting salary, according to the medians provided.

But how do we know $\color{blue}{\text{how many types of schools}}$ are in this dataframe without spending time fishing through the whole set?

Try one of the following two snippets of code:

`dfCollegeType["School Type"].unique()`

$\phantom{xxx}$ The `.unique` function will list the unique qualifiers in the column, or


`set(dfCollegeType["School Type"].values)`

$\phantom{xxx}$ Using both `set` and `values` will give you a set of the values listed in the column.

# Part A

***(1 point)*** Run one of those code snippets in the cell below:



In [14]:
# code snippet here
set(dfCollegeType["School Type"].values)

{'Engineering', 'Ivy League', 'Liberal Arts', 'Party', 'State'}

# Part B
***(1 point)*** **Question: How many school types are there? List them.**

Put your answer to part B here

There are 5 school types in this dataset. They are engineering, ivy league, liberal arts, party, and state schools.
    

### Example help for Part C
If you wanted the mean of the starting median salaries for 'Ivy League' schools then you would apply the `.mean` function to **only** those 'starting median salaries for Ivy League schools.

This can be done with `.loc` as follows:

In [15]:
# Create a variable called `ivy_mean`
ivy_mean = dfCollegeType.loc[dfCollegeType['School Type'] == 'Ivy League', 'Starting Median Salary'].mean()
print("The average (mean) of the starting median salaries for Ivy League schools is ", ivy_mean)


The average (mean) of the starting median salaries for Ivy League schools is  60475.0


# Part C

1] ***(2 points)*** Find the mean of the starting median salaries for **Engineering schools**.

2] ***(2 points)*** Find the mean of the starting median salaries for **State schools**.

3] ***(2 points)*** Find the mean of the starting median salaries for **Liberal Arts schools**.

$\color{red}{\text{(These means need not be edited to dollars and cents from floats)}}$

In [16]:
# SOLUTIONS
#Engineering mean
engineering_mean = dfCollegeType.loc[dfCollegeType['School Type'] == 'Engineering', 'Starting Median Salary'].mean()
print("The mean of starting median salaries for engineering schools is ", engineering_mean)
#State mean
state_mean = dfCollegeType.loc[dfCollegeType['School Type'] == 'State', 'Starting Median Salary'].mean()
print("The mean of starting median salaries for state schools is ", state_mean)
#Liberal Arts Mean
liberal_mean = dfCollegeType.loc[dfCollegeType['School Type'] == 'Liberal Arts', 'Starting Median Salary'].mean()
print("The mean of starting median salaries for liberal arts schools is ", liberal_mean)

The mean of starting median salaries for engineering schools is  59057.89473684211
The mean of starting median salaries for state schools is  44126.28571428572
The mean of starting median salaries for liberal arts schools is  45746.8085106383


# Part D
***(6 points)*** Question: About how much can a person expect to make in the 90th percentile if they are mid-career and got their degree from a party school? 

i.e. report the mean of the mid-career 90th percentile salaries for $\color{red}{\text{party schools}}$.

In [17]:
# Code your solution to part D here
#Cleaning 90th Percentile
dfCollegeType['Mid-Career 90th Percentile Salary'] = dfCollegeType['Mid-Career 90th Percentile Salary'].str.replace(',','')
dfCollegeType['Mid-Career 90th Percentile Salary'] = dfCollegeType['Mid-Career 90th Percentile Salary'].str.replace('$','')
dfCollegeType['Mid-Career 90th Percentile Salary'] = dfCollegeType['Mid-Career 90th Percentile Salary'].astype(float)
#Finding party school mid career salary in 90th percentile
party_90_mean = dfCollegeType.loc[dfCollegeType['School Type'] == 'Party', 'Mid-Career 90th Percentile Salary'].mean(axis = 0, skipna = True)
print("The mean mid-career 90th percentile salary for graduates of party schools is ", party_90_mean)
dfCollegeType

/tmp/ipykernel_378/3027957238.py:4: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  dfCollegeType['Mid-Career 90th Percentile Salary'] = dfCollegeType['Mid-Career 90th Percentile Salary'].str.replace('$','')


The mean mid-career 90th percentile salary for graduates of party schools is  166947.36842105264


School Name  School Type  \
0     Massachusetts Institute of Technology (MIT)  Engineering   
1        California Institute of Technology (CIT)  Engineering   
2                             Harvey Mudd College  Engineering   
3    Polytechnic University of New York, Brooklyn  Engineering   
4                                    Cooper Union  Engineering   
..                                            ...          ...   
264                  Austin Peay State University        State   
265                    Pittsburg State University        State   
266                      Southern Utah University        State   
267           Montana State University - Billings        State   
268                  Black Hills State University        State   

     Starting Median Salary  Mid-Career Median Salary  \
0                   72200.0                  126000.0   
1                   75500.0                  123000.0   
2                   71800.0                  122000.0   
3                   62400.0                  114000.0   
4                   62200.0                  114000.0   
..                      ...                       ...   
264                 37700.0                   59200.0   
265                 40400.0                   58200.0   
266                 41900.0                   56500.0   
267                 37900.0                   50600.0   
268                 35300.0                   43900.0   

     Mid-Career 10th Percentile Salary  Mid-Career 25th Percentile Salary  \
0                              76800.0                            99200.0   
1                                  NaN                           104000.0   
2                                  NaN                            96000.0   
3                              66800.0                            94300.0   
4                                  NaN                            80200.0   
..                                 ...                                ...   
264                            32200.0                            40500.0   
265                            25600.0                            46000.0   
266                            30700.0                            39700.0   
267                            22600.0                            31800.0   
268                            27000.0                            32200.0   

     Mid-Career 75th Percentile Salary  Mid-Career 90th Percentile Salary  
0                             168000.0                           220000.0  
1                             161000.0                                NaN  
2                             180000.0                                NaN  
3                             143000.0                           190000.0  
4                             142000.0                                NaN  
..                                 ...                                ...  
264                            73900.0                            96200.0  
265                            84600.0                           117000.0  
266                            78400.0                           116000.0  
267                            78500.0                            98900.0  
268                            60900.0                            87600.0  

[269 rows x 8 columns]

# Problem 3
Now we would like to use the `.describe` function on our dataframe.

***(3 points)*** Type the following snippet of code into the coding cell below:  `dfCollegeType.describe()`

For full points you want to fully describe what you find in `.describe`. For instance, describe the output of the code and determine if the 5-number summary is included in the output. Why is `count` different for some of the columns? Is `mean` the same as the values calculated above? What is `std`? etc.


In [18]:
# Type the aforementioned code command here
dfCollegeType.describe()

Starting Median Salary  Mid-Career Median Salary  \
count              269.000000                269.000000   
mean             46068.401487              83932.342007   
std               6412.616242              14336.191107   
min              34800.000000              43900.000000   
25%              42000.000000              74000.000000   
50%              44700.000000              81600.000000   
75%              48300.000000              92200.000000   
max              75500.000000             134000.000000   

       Mid-Career 10th Percentile Salary  Mid-Career 25th Percentile Salary  \
count                         231.000000                         269.000000   
mean                        44250.649351                       60373.234201   
std                          8719.612427                       11381.348857   
min                         22600.000000                       31800.000000   
25%                         39000.000000                       53200.000000   
50%                         43100.000000                       58400.000000   
75%                         47400.000000                       65100.000000   
max                         80000.000000                      104000.000000   

       Mid-Career 75th Percentile Salary  Mid-Career 90th Percentile Salary  
count                         269.000000                         231.000000  
mean                       116275.092937                      157705.627706  
std                         22952.334054                       34823.348157  
min                         60900.000000                       87600.000000  
25%                        100000.000000                      136000.000000  
50%                        113000.000000                      153000.000000  
75%                        126000.000000                      170500.000000  
max                        234000.000000                      326000.000000

Put your $\color{red}{\text{written solution}}$ to problem 3 here.

The 5 number summary includes the min and the max numbers, and includes the lower/upper quartiles, and the median. Shown clearly here are the min and max values of the dataset. Furthermore, the 25% ad 75% rows are the 1st and 3rd quartile respectively. The 50% would be the second quartile, also known as the median! Therefore, the output of this code includes the 5 number summary. The count is different in some columns because, as we know from the dataset, some schools didn't include the 10th percentile and 90th percentile statistics, therefore there are less schools to account for in those areas. They are marked on the columns as NaN. All of the mean values match up with their values calculated previously in the lab! Std is the notation for standard deviation, and is a measure of the dispersion of data away from the mean. On symmetrical data sets, roughly 68% of the data will lie within 1 standard deviation of the mean.

# Problem 4

Let $x_1, x_2, \ldots, x_n$ be $n$ observations of a variable of interest.  Recall that the sample mean $\bar{x}$ and sample variance $s^2$ are given by 

$$
\bar{x} = \frac{1}{n}\sum_{k=1}^n x_k \quad \textrm{and} \quad s^2 = \frac{1}{n-1}\sum_{k=1}^n \left( x_k - \bar{x}\right)^2 
$$

Recall standard deviation is $\sqrt{s^2}=s$, i.e. the square root of the variance.

Computationally, $s^2 = \frac{1}{n-1}\sum_{k=1}^n \left( x_k - \bar{x}\right)^2 $ reveals an issue with data manipulation. This formula, when coded, would require two passes over the data: one to compute the mean, and a second to subtract the mean from each observation and compute the sum of squares.

Additionally, this manner of computing sample variance can lead to numerical problems and inefficiencies. For example, if the $x$'s are "large" and the differences between them "small", computing the sample variance using the formula above requires computing a small number as the difference of two small numbers. Not good! Check out this optional wikipedia article about [Loss of Significance](https://en.wikipedia.org/wiki/Loss_of_significance) to learn more.

Therefore, it is useful to be able to compute the variance and/or standard deviation in a single pass, inspecting each value $x_k$ only once; for example, when the data are being collected without enough storage to keep all the values, or when costs of memory access dominate those of computation.
 
We need to manipulate the expressions algebraically in order to arrive at a more computationally efficient formula.

One such identity has been found; it is seen here: 

$$s^2=\frac{1}{n(n-1)} \left(n \cdot \sum_{i=1}^n x_i^2 - \left(\sum_{i=1}^n x_i \right)^2 \right)$$

***(8 points)*** ***Prove*** that the two formulas for standard deviation will in fact calculate the same thing, i.e. prove they are the same formula.

Prove $$\displaystyle s^2 = \frac{\sum_{i=1}^n (x_i-\bar{x})^2}{n-1}$$  and  $$ s^2 = \frac{1}{n(n-1)} \left(n \cdot \sum_{i=1}^n x_i^2 - \left(\sum_{i=1}^n x_i \right)^2 \right)$$ are the same thing.

Solution
$$\displaystyle s^2 = \frac{\sum_{i=1}^n (x_i-\bar{x})^2}{n-1}$$
FOIL the equation inside summation$$\displaystyle= \frac{1}{n-1} (\sum_{i=1}^nx_i^2 - 2 \cdot \bar{x} \sum_{i=1}^nx_i + \sum_{i=1}^n(\bar{x})^2)$$
Using def on mean to solve 2nd summation. 3rd summation has no variable i, so it is multiplied n times$$\displaystyle= \frac{1}{n-1} (\sum_{i=1}^nx_i^2 - 2 \bar{x} \cdot (\bar{x}n) + n(\bar{x})^2)$$
Combining like terms inside parentheses$$\displaystyle= \frac{1}{n-1} (\sum_{i=1}^nx_i^2 - n(\bar{x})^2)$$
Applying def on mean, since x bar is squared we square whole identity$$\displaystyle= \frac{1}{n-1} (\sum_{i=1}^nx_i^2 - n\cdot(\sum_{i=1}^n \frac{x_i}{n})^2)$$
Squaring 1/n, cancelling out numerator n $$\displaystyle= \frac{1}{n-1} (\sum_{i=1}^nx_i^2 - \frac{1}{n}\cdot(\sum_{i=1}^n x_i)^2)$$
Pulling n out $$\displaystyle= \frac{1}{n(n-1)} (n \cdot\sum_{i=1}^nx_i^2 -(\sum_{i=1}^n x_i)^2)$$

QED. Proved two formulas for std deviation are in fact the same